pallets/subtensor/src/subnets/weights.rs

```diff
@@ -35,16 +35,27 @@ impl<T: Config> Pallet<T> {
             Error::<T>::CommitRevealDisabled
         );
 
+        // Check that the validator is not committing too frequently (weights_rate_limit check)
         ensure!(
-            Self::can_commit(netuid, &who),
+            Self::check_rate_limit_for_commit(netuid, &who),
             Error::<T>::WeightsCommitNotAllowed
         );
 
-        WeightCommits::<T>::insert(
-            netuid,
-            &who,
-            (commit_hash, Self::get_current_block_as_u64()),
-        );
+        // Insert the new commit with the current block number.
+        let current_block = Self::get_current_block_as_u64();
+        WeightCommits::<T>::insert(netuid, &who, (commit_hash, current_block));
+
+        // Retain only commits where committed_block < current_block - commit_reveal_interval * 2
+        let interval = Self::get_commit_reveal_weights_interval(netuid).saturating_mul(2);
+        let threshold_block = current_block.saturating_sub(interval);
+
+        // Iterate through the commits and remove old entries
+        WeightCommits::<T>::drain_prefix(netuid).for_each(|(who, (_hash, commit_block))| {
+            if commit_block < threshold_block {
+                WeightCommits::<T>::remove(netuid, who);
```

Comment:
I might be wrong because maybe it's a LinkedList already (which would be another problem), but if it isn't:

Maybe it doesn't matter much because there shouldn't be many of those, but if someone sets the threshold high on a fake subnet, commits a ton of weights from all 64 validators as fast as they can (every 50 blocks I think) and then, say, changes the hyperparameter, it might cause the chain to remove entries from this map inefficiently as soon as someone tries adding something here.

Perhaps it should be a list sorted by block_id and we should bisect in search for the first item on that list to find the first item that is not to be filtered out, and then we'll copy off everything from that point on till the end, making it the new list, causing the entire operation to be log(n) lookups + one memcopy rather than n**2 (looking at each item and removing the (first?) element from an array which requires a rewrite).

Reply:
updated